homeassistant/components/songpal/coordinator.py

```diff
@@ -0,0 +1,259 @@
+
+        async def _source_changed(content: ContentChange):
+            _LOGGER.debug("Source changed: %s", content)
+
+            self.data["inputs"] = await self.device.get_inputs()
+            self.async_set_updated_data(self.data)
+
+        async def _setting_changed(setting: SettingChange):
+            _LOGGER.debug("Setting changed: %s", setting)
+
+            data = self.data
+            data["sound_settings"] = await self.device.get_sound_settings()
+            data["bluetooth_settings"] = await self.device.get_bluetooth_settings()
+            data["misc_settings"] = await self.device.get_misc_settings()
+            data["playback_settings"] = await self.device.get_playback_settings()
+            self.data = data
+            self.async_set_updated_data(self.data)
+
+        async def _power_changed(power: PowerChange):
+            _LOGGER.debug("Power changed: %s", power)
+
+            self.data["power"] = await self.device.get_power()
+            self.async_set_updated_data(self.data)
+
+        async def _try_reconnect(connect: ConnectChange):
+            _LOGGER.warning(
+                "[%s(%s)] Got disconnected, trying to reconnect (%s)",
+                self.name,
+                self.device.endpoint,
+                connect.exception,
+            )
+            self.available = False
+            self.async_set_updated_data(self.data)
+
+            # Try to reconnect forever, a successful reconnect will initialize
+            # the websocket connection again.
+            delay = INITIAL_RETRY_DELAY
+            while not self.available:
+                _LOGGER.debug("Trying to reconnect in %s seconds", delay)
+                await asyncio.sleep(delay)
+
+                try:
+                    await self.device.get_supported_methods()
+                except SongpalException as ex:
+                    _LOGGER.debug("Failed to reconnect: %s", ex)
+                    delay = min(2 * delay, 300)
+                else:
+                    # We need to inform HA about the state in case we are coming
+                    # back from a disconnected state.
+                    await self.full_refresh()
+                    self.async_set_updated_data(self.data)
+
+            self.hass.loop.create_task(self.device.listen_notifications())
+            _LOGGER.info(
+                "[%s(%s)] Connection reestablished", self.name, self.device.endpoint
+            )
+
+        self.device.on_notification(VolumeChange, _volume_changed)
+        self.device.on_notification(ContentChange, _source_changed)
+        self.device.on_notification(PowerChange, _power_changed)
+        self.device.on_notification(SettingChange, _setting_changed)
+        self.device.on_notification(ConnectChange, _try_reconnect)
+
+        async def handle_stop(event):
+            await self.device.stop_listen_notifications()
+
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, handle_stop)
```

Comment:
We need to unload this listener once the entry is unloaded. Use entry.async_on_unload for this

Idem for the handle_stop btw